### PR TITLE
Kafka components

### PR DESCRIPTION
**Pull request by erykk8**
Consumers (Email, Slack) - adaptors for email and slack sending, consume notifications from Kafka and send them out
Producer - fetches weather data and publishes it to Kafka
Stream - alert-flow component, takes in weather data, gets alerts from database, detects if conditions were met, saves occurrences in the database and sends out notifications

Email and Slack classes were simplified to not use the Akka actor system

### CODE REVIEW

File: src/main/scala/notifications/kafka/Producer.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  def produceLocationData(location: String) = {
+    val weather = getLocationData(location)
+    val data = new ProducerRecord[String, Weather](config.topic, location, weather)
+    producer.send(data)
+  }
+
+  override def run(): Unit = {
+    try {
+      while (true) {
```
Comment (RobPiwowarek) [inline src/main/scala/notifications/kafka/Producer.scala]: I suppose there is no scala alternative to doing this?

File: src/main/scala/notifications/kafka/Producer.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        // send out weather data for active locations, location name being the message key
+        for (location <- getActiveLocations()) {
+          produceLocationData(location)
+        }
+        Thread.sleep(10 * 60 * 1000) // every 10 minutes
```
Comment (RobPiwowarek) [inline src/main/scala/notifications/kafka/Producer.scala]: you could import scala.concurrent.duration._ and do smthing like Thread.sleep(10.minutes.length) technically should work the same
also move this to an external parameter or config

File: src/main/scala/notifications/kafka/Serdes.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  object Slack extends ScalaSerde[SlackNotification] with JsonSupport {
+
+    final object Deserializer extends Deserializer[SlackNotification] {
+      override def deserialize(data: Array[Byte]): Option[SlackNotification] = {
+        try {
+          val jsonData = new String(data).parseJson
+          Some(jsonData.convertTo[SlackNotification])
+        }
+        catch {
+          case e: ParsingException => {
```
Comment (RobPiwowarek) [inline src/main/scala/notifications/kafka/Serdes.scala]: inconsistent { } with Email's one

File: src/main/scala/notifications/kafka/Serdes.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  object Email extends ScalaSerde[EmailNotification] with JsonSupport {
+
+    final object Deserializer extends Deserializer[EmailNotification] {
+      override def deserialize(data: Array[Byte]): Option[EmailNotification] = {
+        try {
+          val jsonData = new String(data).parseJson
+          Some(jsonData.convertTo[EmailNotification])
+        }
+        catch {
+          case e: ParsingException => None
```
Comment (RobPiwowarek) [inline src/main/scala/notifications/kafka/Serdes.scala]: is this really something we want? completely ignoring exceptions?

File: src/main/scala/notifications/kafka/Serdes.scala [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  object Weather extends ScalaSerde[Responses.Weather] with JsonSupport {
+
+    final object Deserializer extends Deserializer[Responses.Weather] {
+      override def deserialize(data: Array[Byte]): Option[Responses.Weather] = {
+        try {
+          val jsonData = new String(data).parseJson
+          Some(jsonData.convertTo[Responses.Weather])
+        }
+        catch {
+          case e: ParsingException => {
```
Comment (RobPiwowarek) [inline src/main/scala/notifications/kafka/Serdes.scala]: inconsistent { } with Email's one

File: src/main/scala/notifications/kafka/Stream.scala [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      }
+      case "HUMI" => weather.main.humidity
+      case "PRES" => weather.main.pressure
+      case "CLOU" => BigDecimal(weather.clouds.all)
+    }
+  }
+
+  def conditionIsMet(param: DefinitionParameter, weather: Weather): Boolean = {
+    val value = getValue(param, weather)
+    param.comparisonType match {
```
Comment (RobPiwowarek) [inline src/main/scala/notifications/kafka/Stream.scala]: same as above, match _ to some throw/message

File: src/main/scala/notifications/kafka/Stream.scala [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  override def run(): Unit = {
+    val streams = createStream()
+    streams.start()
+
+    Runtime.getRuntime.addShutdownHook(new Thread {
+      override def run(): Unit = {
+        try {
+          streams.close()
+        }
+        finally {
```
Comment (RobPiwowarek) [inline src/main/scala/notifications/kafka/Stream.scala]: this achieves nothing

**Review by RobPiwowarek — CHANGES_REQUESTED** [review]
hit intellij's reformat code and optimize imports on all the files or recursively on a directory, also change case classes that we dont extend to final case classes

File: src/main/scala/notifications/Notification.scala [hunk 8]
```diff
@@ -1,24 +1,55 @@
...
+
 
-case class SlackNotification(override val user: String,
-                             slackUsername: String,
-                             override val alert: String) extends Notification {}
\ No newline at end of file
+  override def toString = {
+    val alertString = Helpers.alertParamToString(location, parameters)
+    s"""
+       | Hi $slackUsername!\n\n
```
[the remark this answers is not in the document]
Comment (agatak8) [reply src/main/scala/notifications/Notification.scala]: they did not output "\n\n" when i tested this

[the remark this answers is not in the document]
Comment (RobPiwowarek) [reply src/main/scala/notifications/Notification.scala]: so they were actually new lines?

[the remark this answers is not in the document]
Comment (agatak8) [reply src/main/scala/notifications/Notification.scala]: Yes

File: src/main/scala/notifications/kafka/Serdes.scala (hunk 4 above)
Comment (agatak8) [reply src/main/scala/notifications/kafka/Serdes.scala]: no, it should at least log this somewhere but do we have the time

File: src/main/scala/notifications/kafka/Producer.scala (hunk 1 above)
Comment (agatak8) [reply src/main/scala/notifications/kafka/Producer.scala]: you mean using something else than Threads? should we maybe run them as separate jars?

File: src/main/scala/notifications/Notification.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (RobPiwowarek) [reply src/main/scala/notifications/Notification.scala]: i m losing faith in raw strings

File: src/main/scala/notifications/kafka/Producer.scala (hunk 1 above)
Comment (RobPiwowarek) [reply src/main/scala/notifications/kafka/Producer.scala]: I meant not using while(true) and try {}. I hardly ever see that used in scala

File: src/main/scala/notifications/kafka/Configs.scala [hunk 9]
```diff
@@ -93,6 +96,7 @@ object Configs {
   }
 
   object Producer extends WeatherProducerConfig {
+    val delay = Duration(10, MINUTES)
```
Comment (RobPiwowarek) [inline src/main/scala/notifications/kafka/Configs.scala]: i literally meant you could import scala.concurrent.duration._ and use "10.minutes"

**Review by RobPiwowarek — COMMENTED** [review]
merge if you don't care about that one comment that is left for you

**Review by RobPiwowarek — APPROVED** [review]

File: src/main/scala/notifications/kafka/Configs.scala (hunk 9 above)
Comment (agatak8) [reply src/main/scala/notifications/kafka/Configs.scala]: it didn't show up in IntelliJ so I took the safe route

**Pull request merged**